Source/Dafny/AST/DafnyAst.cs

```diff
@@ -2093,6 +2090,18 @@ public enum AutoInitInfo { MaybeEmpty, Nonempty, CompilableValue }
       Contract.Requires(b != null);
       Contract.Requires(builtIns != null);
 
+      a = a.NormalizeExpandKeepConstraints();
+      b = b.NormalizeExpandKeepConstraints();
+      if (a is IntVarietiesSupertype) {
+        return b is IntVarietiesSupertype || b.IsNumericBased(NumericPersuasion.Int) || b.IsBigOrdinalType || b.IsBitVectorType ? b : null;
+      } else if (b is IntVarietiesSupertype) {
+        return a.IsNumericBased(NumericPersuasion.Int) || a.IsBigOrdinalType || a.IsBitVectorType ? a : null;
+      } else if (a is RealVarietiesSupertype) {
+        return b is RealVarietiesSupertype || b.IsNumericBased(NumericPersuasion.Real) ? b : null;
+      } else if (b is RealVarietiesSupertype) {
+        return a.IsNumericBased(NumericPersuasion.Real) ? a : null;
```

Comment:
Nitpick: I find the asymmetry here confusing.  Is it semantically relevant that the check on `a` goes first?  I'm not fluent in what the two variants (`x is XVarietiesSupertype` vs. `x.IsNumericaBased(NumericPersuasion.X)`) mean, but I wonder if there would be a more symmetrical way to phrase this check.

Reply:
I'm leaving this as is. The identity of values of type `RealVarietiesSupertype` [or, analogously, `IntVarietiesSupertype`] is not important. These types just say "this is a placeholder that stands for any real-valued [int-valued, respectively] type". So, the only asymmetry is that if both `a` and `b` satisfy `is RealVarietiesSupertype`, then `b` (not `a`) is returned, but this choice is arbitrary.